Null pointer in tablet servers  compaction metrics code.

**Describe the bug**

Observed the following stack trace in tablet server logs.  


```
2025-02-24T23:05:56,261 [threads.ThreadPools] ERROR: Non-critical scheduled background task failed
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "java.util.concurrent.atomic.AtomicInteger.set(int)" because "cem.running" is null
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[?:?]
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191) ~[?:?]
        at org.apache.accumulo.core.util.threads.ThreadPools.checkTaskFailed(ThreadPools.java:151) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at org.apache.accumulo.core.util.threads.ThreadPools$1.lambda$run$0(ThreadPools.java:120) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at org.apache.accumulo.core.util.threads.ThreadPools$1.run(ThreadPools.java:117) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at org.apache.accumulo.core.trace.TraceWrappedRunnable.run(TraceWrappedRunnable.java:52) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at org.apache.accumulo.core.trace.TraceWrappedRunnable.run(TraceWrappedRunnable.java:52) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.concurrent.atomic.AtomicInteger.set(int)" because "cem.running" is null
        at org.apache.accumulo.tserver.metrics.CompactionExecutorsMetrics.lambda$update$4(CompactionExecutorsMetrics.java:148) ~[accumulo-tserver-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at org.apache.accumulo.tserver.metrics.CompactionExecutorsMetrics.update(CompactionExecutorsMetrics.java:147) ~[accumulo-tserver-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at org.apache.accumulo.core.trace.TraceWrappedRunnable.run(TraceWrappedRunnable.java:52) ~[accumulo-core-2.1.4-SNAPSHOT.jar:2.1.4-SNAPSHOT]
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) ~[?:?]
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
        ... 4 more
```

--- a/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
+++ b/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
@@ -61,12 +61,8 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
     public void close() {
       runningSupplier = () -> 0;
       queuedSupplier = () -> 0;
-      if (running != null) {
-        running.set(0);
-      }
-      if (queued != null) {
-        queued.set(0);
-      }
+      running.set(0);
+      queued.set(0);
     }
   }
 
@@ -135,34 +131,22 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
             return m;
           });
 
-          if (exm.queued != null) {
-            exm.queued.set(ecm.queued);
-          }
-          if (exm.running != null) {
-            exm.running.set(ecm.running);
-          }
+          exm.queued.set(ecm.queued);
+          exm.running.set(ecm.running);
 
         });
 
         Sets.difference(exCeMetricsMap.keySet(), seenIds).forEach(unusedId -> {
           ExMetrics exm = exCeMetricsMap.get(unusedId);
-          if (exm.queued != null) {
-            exm.queued.set(0);
-          }
-          if (exm.running != null) {
-            exm.running.set(0);
-          }
+          exm.queued.set(0);
+          exm.running.set(0);
         });
 
       }
     }
     ceMetricsList.forEach(cem -> {
-      if (cem.running != null) {
-        cem.running.set(cem.runningSupplier.getAsInt());
-      }
-      if (cem.queued != null) {
-        cem.queued.set(cem.queuedSupplier.getAsInt());
-      }
+      cem.running.set(cem.runningSupplier.getAsInt());
+      cem.queued.set(cem.queuedSupplier.getAsInt());
     });
   }
 

--- a/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
+++ b/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
@@ -61,8 +61,12 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
     public void close() {
       runningSupplier = () -> 0;
       queuedSupplier = () -> 0;
-      running.set(0);
-      queued.set(0);
+      if (running != null) {
+        running.set(0);
+      }
+      if (queued != null) {
+        queued.set(0);
+      }
     }
   }
 
@@ -131,22 +135,34 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
             return m;
           });
 
-          exm.queued.set(ecm.queued);
-          exm.running.set(ecm.running);
+          if (exm.queued != null) {
+            exm.queued.set(ecm.queued);
+          }
+          if (exm.running != null) {
+            exm.running.set(ecm.running);
+          }
 
         });
 
         Sets.difference(exCeMetricsMap.keySet(), seenIds).forEach(unusedId -> {
           ExMetrics exm = exCeMetricsMap.get(unusedId);
-          exm.queued.set(0);
-          exm.running.set(0);
+          if (exm.queued != null) {
+            exm.queued.set(0);
+          }
+          if (exm.running != null) {
+            exm.running.set(0);
+          }
         });
 
       }
     }
     ceMetricsList.forEach(cem -> {
-      cem.running.set(cem.runningSupplier.getAsInt());
-      cem.queued.set(cem.queuedSupplier.getAsInt());
+      if (cem.running != null) {
+        cem.running.set(cem.runningSupplier.getAsInt());
+      }
+      if (cem.queued != null) {
+        cem.queued.set(cem.queuedSupplier.getAsInt());
+      }
     });
   }
 

--- a/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
+++ b/server/tserver/src/main/java/org/apache/accumulo/tserver/metrics/CompactionExecutorsMetrics.java
@@ -50,12 +50,25 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
   private MeterRegistry registry = null;
 
   // public so it can be closed by outside callers
-  public static class CeMetrics implements AutoCloseable {
-    private AtomicInteger queued;
-    private AtomicInteger running;
+  public class CeMetrics implements AutoCloseable {
+    private final AtomicInteger queued;
+    private final AtomicInteger running;
 
     private IntSupplier runningSupplier;
     private IntSupplier queuedSupplier;
+
+    private CeMetrics(CompactionExecutorId ceid) {
+      if (registry != null) {
+        this.queued = registry.gauge(METRICS_MAJC_QUEUED, Tags.of("id", ceid.canonical()),
+            new AtomicInteger(0));
+        this.running = registry.gauge(METRICS_MAJC_RUNNING, Tags.of("id", ceid.canonical()),
+            new AtomicInteger(0));
+      } else {
+        // these vars have no effect on metrics in this case - just avoids NPEs
+        this.queued = new AtomicInteger(0);
+        this.running = new AtomicInteger(0);
+      }
+    }
 
     @Override
     public void close() {
@@ -66,9 +79,22 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
     }
   }
 
-  private static class ExMetrics {
-    AtomicInteger queued;
-    AtomicInteger running;
+  private class ExMetrics {
+    private final AtomicInteger queued;
+    private final AtomicInteger running;
+
+    private ExMetrics(CompactionExecutorId ceid) {
+      if (registry != null) {
+        this.queued = registry.gauge(METRICS_MAJC_QUEUED, Tags.of("id", ceid.canonical()),
+            new AtomicInteger(0));
+        this.running = registry.gauge(METRICS_MAJC_RUNNING, Tags.of("id", ceid.canonical()),
+            new AtomicInteger(0));
+      } else {
+        // these vars have no effect on metrics in this case - just avoids NPEs
+        this.queued = new AtomicInteger(0);
+        this.running = new AtomicInteger(0);
+      }
+    }
   }
 
   public CompactionExecutorsMetrics() {
@@ -89,16 +115,7 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
 
     synchronized (ceMetricsMap) {
 
-      CeMetrics cem = ceMetricsMap.computeIfAbsent(ceid, id -> {
-        CeMetrics m = new CeMetrics();
-        if (registry != null) {
-          m.queued = registry.gauge(METRICS_MAJC_QUEUED, Tags.of("id", ceid.canonical()),
-              new AtomicInteger(0));
-          m.running = registry.gauge(METRICS_MAJC_RUNNING, Tags.of("id", ceid.canonical()),
-              new AtomicInteger(0));
-        }
-        return m;
-      });
+      CeMetrics cem = ceMetricsMap.computeIfAbsent(ceid, id -> new CeMetrics(ceid));
 
       cem.runningSupplier = runningSupplier;
       cem.queuedSupplier = queuedSupplier;
@@ -120,16 +137,7 @@ public class CompactionExecutorsMetrics implements MetricsProducer {
         externalMetricsSupplier.get().forEach(ecm -> {
           seenIds.add(ecm.ceid);
 
-          ExMetrics exm = exCeMetricsMap.computeIfAbsent(ecm.ceid, id -> {
-            ExMetrics m = new ExMetrics();
-            if (registry != null) {
-              m.queued = registry.gauge(METRICS_MAJC_QUEUED, Tags.of("id", ecm.ceid.canonical()),
-                  new AtomicInteger(0));
-              m.running = registry.gauge(METRICS_MAJC_RUNNING, Tags.of("id", ecm.ceid.canonical()),
-                  new AtomicInteger(0));
-            }
-            return m;
-          });
+          ExMetrics exm = exCeMetricsMap.computeIfAbsent(ecm.ceid, id -> new ExMetrics(ecm.ceid));
 
           exm.queued.set(ecm.queued);
           exm.running.set(ecm.running);

